diem-move/parallel-executor/src/scheduler.rs

```diff
@@ -1,113 +1,469 @@
-            execution_marker: AtomicUsize::new(0),
-            stop_at_version: AtomicUsize::new(num_txns),
-            txn_buffer: SegQueue::new(),
+            execution_idx: AtomicUsize::new(0),
+            validation_idx: AtomicUsize::new(0),
+            decrease_cnt: AtomicUsize::new(0),
+            num_active_tasks: AtomicUsize::new(0),
+            done_marker: AtomicBool::new(false),
+            stop_idx: AtomicUsize::new(num_txns),
+            block_size: num_txns,
             txn_dependency: (0..num_txns)
-                .map(|_| Arc::new(RwLock::new(Vec::new())))
+                .map(|_| CachePadded::new(Mutex::new(Vec::new())))
                 .collect(),
             txn_status: (0..num_txns)
-                .map(|_| AtomicUsize::new(ExecutionStatus::NotExecuted as usize))
+                .map(|_| CachePadded::new(Mutex::new(TransactionStatus::ReadyToExecute(0))))
                 .collect(),
         }
     }
 
-    // Return the next txn id for the thread to execute: first fetch from the shared queue that
-    // stores dependency-resolved txns, then fetch from the original ordered txn sequence.
-    // Return Some(id) if found the next transaction, else return None.
-    pub fn next_txn_to_execute(&self) -> Option<Version> {
-        // Fetch txn from txn_buffer
-        match self.txn_buffer.pop() {
-            Some(version) => Some(version),
-            None => {
-                // Fetch the first non-executed txn from the original transaction list
-                let next_to_execute = self.execution_marker.fetch_add(1, Ordering::Relaxed);
-                if next_to_execute < self.num_txn_to_execute() {
-                    Some(next_to_execute)
-                } else {
-                    // Everything executed at least once - validation will take care of rest.
-                    None
+    // Decreases the validation index, increases the decrease counter if it actually decreased.
+    fn decrease_validation_idx(&self, target_idx: TxnIndex) {
+        if self.validation_idx.fetch_min(target_idx, Ordering::SeqCst) > target_idx {
+            self.decrease_cnt.fetch_add(1, Ordering::SeqCst);
+        }
+    }
+
+    // Decreases the execution index, increases the decrease counter if it actually decreased.
+    fn decrease_execution_idx(&self, target_idx: TxnIndex) {
+        if self.execution_idx.fetch_min(target_idx, Ordering::SeqCst) > target_idx {
+            self.decrease_cnt.fetch_add(1, Ordering::SeqCst);
+        }
+    }
+
+    // Try to abort version = (txn_idx, incarnation), called upon validation failure.
+    // When the invocation manages to update the status of the transaction
+    // Executed(incarnation) => Aborting(incarnation), it returns true. Otherwise,
+    // returns false. Since incarnation numbers never decrease, this also ensures
+    // that the same version may not successfully abort more than once.
+    pub fn abort(&self, txn_idx: TxnIndex, incarnation: Incarnation) -> bool {
```

Comment:
one thing I'd like is to have separate impl block for public and private functions so it's more explicit for readers which functions are public interface.
/// public interfaces
impl Scheduler {
  pub fn f1...
  pub fn f2...
  ...
}
/// private functions
impl Scheduler {
  fn f3..
  fn f4..
   ...
}